service/r2/service.go

```diff
@@ -192,6 +193,7 @@ func (s *service) RegisterHandlers(router *mux.Router) {
 
 	router.Handle(namespacePath, h.wrap(s.fetchNamespaces)).Methods(http.MethodGet)
 	router.Handle(namespacePath, h.wrap(s.createNamespace)).Methods(http.MethodPost)
+	router.Handle(namespaceValidatePath, h.wrap(s.validateNamespace)).Methods(http.MethodPost)
```

Comment:
How is auth going to work for validation endpoint? Is it "if a user has read access, then he/she can validate?" Or does he/she need to have write access to validate?

Reply:
Left a comment on this earlier - not sure why it didn't post. In any case this is a good point. If the requester does not have write access, due to how the current auth is set up, they will not be validated for the endpoint. The way around this is to make the endpoint GET or to change how we do auth.

Reply:
I think we can make the `NewAuthHandler` take an additional parameter `authType`:
```
type authorizationType int

const (
  none authorizationType
  readOnly
  writeOnly
  readWrite
)

func (a simpleAuth) NewAuthHandler(
  authType authorizationType,
  next http.Handler,
  errHandler errorResponseHandler,
) http.Handler {
  ...
}

return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
  ...
  switch authType {
    case none:
      ...
    case readOnly:
      ...
    case writeOnly:
      ...
    case readWrite:
      ...
}
```
Then in `service.go`, register the auth type
```
func registerRoute(router *mux.Router, path, method string, handler http.Handler) error {
  authType, exists := authRegistry[endpoint{Path:path, Method: method}]
  if !exists {
    // Should make this a method instead.
    switch method:
    case http.GET:
      authType = readOnly
    case http.POST, ...:
      authType = writeOnly
    default:
      return errUnknownHTTPMethod
  }
  fn := h.wrap(authType, handler)
  router.Handle(path, fn).Methods(method)
  return nil
}

func (h r2Handler) wrap(authType authorizationType, fn r2HandlerFunc) http.Handler {
	f := http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		if err := fn(w, r); err != nil {
			h.handleError(w, err)
		}
	})
	return h.auth.NewAuthHandler(authType, f, writeAPIResponse)
}

type endpoint struct {
  Path string
  Method string
}

var authRegistry = map[endpoint]authorizationType{
  {Path: namespaceValidePath, Method: http.POST}: readOnly
}
```